packages/editor-ui/src/mixins/pinData.ts

```diff
@@ -83,22 +95,104 @@ export const pinData = defineComponent({
 				return false;
 			}
 		},
-		isValidPinDataSize(data: string | object): boolean {
+		isValidPinDataSize(data: string | object, activeNodeName: string): boolean {
 			if (typeof data === 'object') data = JSON.stringify(data);
 
-			if (
-				this.workflowsStore.pinDataSize + stringSizeInBytes(data) >
-				MAX_WORKFLOW_PINNED_DATA_SIZE
-			) {
+			const { pinData: currentPinData, ...workflow } = this.workflowsStore.getCurrentWorkflow();
+			const workflowJson = jsonStringify(workflow, { replaceCircularRefs: true });
+
+			const newPinData = { ...currentPinData, [activeNodeName]: data };
+			const newPinDataSize = this.workflowsStore.getPinDataSize(newPinData);
+
+			let isValid = true;
+			if (newPinDataSize > MAX_WORKFLOW_PIN_DATA_SIZE) {
 				this.showError(
 					new Error(this.$locale.baseText('ndv.pinData.error.tooLarge.description')),
 					this.$locale.baseText('ndv.pinData.error.tooLarge.title'),
 				);
 
+				isValid = false;
+			}
+
+			if (
+				stringSizeInBytes(workflowJson) + newPinDataSize >
+				MAX_WORKFLOW_SIZE - MAX_EXPECTED_REQUEST_SIZE
+			) {
+				this.showError(
+					new Error(this.$locale.baseText('ndv.pinData.error.tooLargeWorkflow.description')),
+					this.$locale.baseText('ndv.pinData.error.tooLargeWorkflow.title'),
+				);
+
+				isValid = false;
+			}
+
+			return isValid;
+		},
+		setPinData(node: INodeUi | null, data: string | INodeExecutionData[], source: string): boolean {
```

Comment:
1. Node will never be null, changed definition and added checks
2. I've changed the flow to throw an error instead of boolean and using try/catch instead. 
3. Unfortunately can't. It's inside a mixin and it's using data from the component it's used in.....